platform/andes/ae350/plmt.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2019 Andes Technology Corporation
+ *
+ * Authors:
+ *   Zong Li <zong@andestech.com>
+ *   Nylon Chen <zong@andestech.com>
+ */
+
+#include <sbi/riscv_io.h>
+#include <sbi/sbi_hart.h>
+
+static u32 plmt_time_hart_count;
+static volatile void *plmt_time_base;
+static volatile u64 *plmt_time_val;
+static volatile u64 *plmt_time_cmp;
+
+u64 plmt_timer_value(void)
+{
+#if __riscv_xlen == 64
+	return readq_relaxed(plmt_time_val);
+#else
+	u64 tmp;
+
+	tmp = readl_relaxed((void *)plmt_time_val + 0x04);
+	tmp <<= 32;
+	tmp |= readl_relaxed(plmt_time_val);
+	return tmp;
```

Comment:
Reading 64bit free running MMIO timer can be tricky on 32bit system because you have to consider higher 32bits can change by the time we read lower 32bits.

This can be addressed using "do { } while" as follows:

```
u32 lo, hi;
do {
    hi = readl_relaxed((void *)plmt_time_val + 0x04);
    lo = readl_relaxed(plmt_time_val);
} while (hi != readl_relaxed((void *)plmt_time_val + 0x04));

return( (u64)hi << 32) | (u64)lo;

```

Reply:
This needs to be fixed in clint_timer_value() as well.